File: golang/direct_e2ee/v2_models.go
package directe2ee

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"strings"
	"time"
)

const (
	DirectE2EEProfileV2               = "anp.direct.e2ee.v2"
	DirectE2EESecurityProfile         = "direct-e2ee"
	TransportProtectedSecurityProfile = "transport-protected"
	ContentTypeDirectInitV2           = "application/anp-direct-init+json"
	ContentTypeDirectCipherV2         = "application/anp-direct-cipher+json"
	MTIDirectE2EESuiteV2              = "ANP-DIRECT-E2EE-X3DH-25519-CHACHA20POLY1305-SHA256-V1"
)

type V2SignedPrekey struct {
	KeyID         string `json:"key_id"`
	PublicKeyB64U string `json:"public_key_b64u"`
	ExpiresAt     string `json:"expires_at"`
}

func (value V2SignedPrekey) Validate() error {
	if value.KeyID == "" {
		return invalidV2("signed_prekey.key_id must be non-empty")
	}
	if err := validateX25519B64U(value.PublicKeyB64U, "signed_prekey.public_key_b64u"); err != nil {
		return err
	}
	if _, err := time.Parse(time.RFC3339, value.ExpiresAt); err != nil {
		return invalidV2("signed_prekey.expires_at must be RFC3339")
	}
	return nil
}

type V2OneTimePrekey struct {
	KeyID         string `json:"key_id"`
	PublicKeyB64U string `json:"public_key_b64u"`
}

func (value V2OneTimePrekey) Validate() error {
	if value.KeyID == "" {
		return invalidV2("one_time_prekey.key_id must be non-empty")
	}
	return validateX25519B64U(value.PublicKeyB64U, "one_time_prekey.public_key_b64u")
}

type V2PrekeyBundle struct {
	BundleID             string         `json:"bundle_id"`
	OwnerDID             string         `json:"owner_did"`
	OwnerDeviceID        string         `json:"owner_device_id"`
	Suite                string         `json:"suite"`
	StaticKeyAgreementID string         `json:"static_key_agreement_id"`
	SignedPrekey         V2SignedPrekey `json:"signed_prekey"`
	Proof                map[string]any `json:"proof"`
}

func (value V2PrekeyBundle) ValidateStructure() error {
	if empty(value.BundleID, value.OwnerDID, value.OwnerDeviceID, value.StaticKeyAgreementID) {
		return invalidV2("bundle identifiers must be non-empty")
	}
	if value.Suite != MTIDirectE2EESuiteV2 {
		return invalidV2("unsupported suite")
	}
	if err := value.SignedPrekey.Validate(); err != nil {
		return err
	}
	for _, field := range []string{"type", "cryptosuite", "verificationMethod", "proofPurpose", "created", "proofValue"} {
		if stringValue(value.Proof[field]) == "" {
			return invalidV2("missing bundle proof field: " + field)
		}
	}
	if stringValue(value.Proof["type"]) != "DataIntegrityProof" || stringValue(value.Proof["cryptosuite"]) != "eddsa-jcs-2022" || stringValue(value.Proof["proofPurpose"]) != "assertionMethod" {
		return invalidV2("invalid Appendix-B Object Proof profile")
	}
	return nil
}

type V2Target struct {
	Kind string `json:"kind"`
	DID  string `json:"did"`
}

type V2KeyServiceMetadata struct {
	ANPVersion      string   `json:"anp_version,omitempty"`
	Profile         string   `json:"profile"`
	SecurityProfile string   `json:"security_profile"`
	SenderDID       string   `json:"sender_did"`
	SenderDeviceID  string   `json:"sender_device_id"`
	Target          V2Target `json:"target"`
	OperationID     string   `json:"operation_id"`
	CreatedAt       string   `json:"created_at,omitempty"`
}

func (value V2KeyServiceMetadata) Validate() error {
	if value.Profile != DirectE2EEProfileV2 || value.SecurityProfile != TransportProtectedSecurityProfile {
		return invalidV2("invalid key-service profile binding")
	}
	if value.Target.Kind != "service" {
		return invalidV2("key-service target.kind must be service")
	}
	if empty(value.SenderDID, value.SenderDeviceID, value.Target.DID, value.OperationID) {
		return invalidV2("key-service selectors must be non-empty")
	}
	if value.CreatedAt != "" {
		if _, err := time.Parse(time.RFC3339, value.CreatedAt); err != nil {
			return invalidV2("meta.created_at must be RFC3339")
		}
	}
	return nil
}

type V2DirectMetadata struct {
	ANPVersion        string   `json:"anp_version,omitempty"`
	Profile           string   `json:"profile"`
	SecurityProfile   string   `json:"security_profile"`
	SenderDID         string   `json:"sender_did"`
	SenderDeviceID    string   `json:"sender_device_id"`
	Target            V2Target `json:"target"`
	RecipientDeviceID string   `json:"recipient_device_id"`
	OperationID       string   `json:"operation_id"`
	MessageID         string   `json:"message_id"`
	ContentType       string   `json:"content_type"`
	CreatedAt         string   `json:"created_at,omitempty"`
}

func (value V2DirectMetadata) Validate() error {
	if value.Profile != DirectE2EEProfileV2 || value.SecurityProfile != DirectE2EESecurityProfile {
		return invalidV2("invalid direct.send profile binding")
	}
	if value.Target.Kind != "agent" {
		return invalidV2("direct.send target.kind must be agent")
	}
	if value.ContentType != ContentTypeDirectInitV2 && value.ContentType != ContentTypeDirectCipherV2 {
		return invalidV2("content_type is not a P5 v2 MTI wire object")
	}
	if empty(value.SenderDID, value.SenderDeviceID, value.Target.DID, value.RecipientDeviceID, value.OperationID, value.MessageID) {
		return invalidV2("direct.send selectors and IDs must be non-empty")
	}
	if value.OperationID != value.MessageID {
		return invalidV2("operation_id must equal message_id")
	}
	if value.CreatedAt != "" {
		if _, err := time.Parse(time.RFC3339, value.CreatedAt); err != nil {
			return invalidV2("meta.created_at must be RFC3339")
		}
	}
	return nil
}

type V2RatchetHeader struct {
	DHPubB64U string `json:"dh_pub_b64u"`
	PN        string `json:"pn"`
	N         string `json:"n"`
}

func (value V2RatchetHeader) Validate() error {
	if err := validateX25519B64U(value.DHPubB64U, "ratchet_header.dh_pub_b64u"); err != nil {
		return err
	}
	if empty(value.PN, value.N) || !decimal(value.PN) || !decimal(value.N) {
		return invalidV2("invalid ratchet header")
	}
	return nil
}

type V2DirectInitBody struct {
	SessionID                  string `json:"session_id"`
	Suite                      string `json:"suite"`
	SenderStaticKeyAgreementID string `json:"sender_static_key_agreement_id"`
	RecipientBundleID          string `json:"recipient_bundle_id"`
	RecipientSignedPrekeyID    string `json:"recipient_signed_prekey_id"`
	RecipientOneTimePrekeyID   string `json:"recipient_one_time_prekey_id,omitempty"`
	SenderEphemeralPubB64U     string `json:"sender_ephemeral_pub_b64u"`
	CiphertextB64U             string `json:"ciphertext_b64u"`
}

func (value V2DirectInitBody) Validate() error {
	if value.Suite != MTIDirectE2EESuiteV2 {
		return invalidV2("unsupported suite")
	}
	if err := validateFixedB64U(value.SessionID, "body.session_id", 16); err != nil {
		return err
	}
	if err := validateX25519B64U(value.SenderEphemeralPubB64U, "body.sender_ephemeral_pub_b64u"); err != nil {
		return err
	}
	if err := validateB64U(value.CiphertextB64U, "body.ciphertext_b64u"); err != nil {
		return err
	}
	if empty(value.SenderStaticKeyAgreementID, value.RecipientBundleID, value.RecipientSignedPrekeyID) {
		return invalidV2("init fields must be non-empty")
	}
	return nil
}

type V2DirectCipherBody struct {
	SessionID      string          `json:"session_id"`
	Suite          string          `json:"suite,omitempty"`
	RatchetHeader  V2RatchetHeader `json:"ratchet_header"`
	CiphertextB64U string          `json:"ciphertext_b64u"`
}

func (value V2DirectCipherBody) Validate() error {
	if err := validateFixedB64U(value.SessionID, "body.session_id", 16); err != nil {
		return err
	}
	if err := validateB64U(value.CiphertextB64U, "body.ciphertext_b64u"); err != nil {
		return err
	}
	if value.Suite != "" && value.Suite != MTIDirectE2EESuiteV2 {
		return invalidV2("cipher suite mismatch")
	}
	return value.RatchetHeader.Validate()
}

type V2ApplicationPlaintext struct {
	ApplicationContentType string          `json:"application_content_type"`
	LogicalMessageID       string          `json:"logical_message_id,omitempty"`
	ConversationID         string          `json:"conversation_id,omitempty"`
	ReplyToMessageID       string          `json:"reply_to_message_id,omitempty"`
	Annotations            *map[string]any `json:"annotations,omitempty"`
	Text                   *string         `json:"text,omitempty"`
	Payload                *map[string]any `json:"payload,omitempty"`
	PayloadB64U            *string         `json:"payload_b64u,omitempty"`
}

func (value *V2ApplicationPlaintext) UnmarshalJSON(encoded []byte) error {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(encoded, &raw); err != nil {
		return invalidV2("invalid ApplicationPlaintext")
	}
	for _, field := range []string{
		"logical_message_id", "conversation_id", "reply_to_message_id",
		"annotations", "text", "payload", "payload_b64u",
	} {
		if member, present := raw[field]; present && bytes.Equal(bytes.TrimSpace(member), []byte("null")) {
			return invalidV2(field + " must be omitted rather than null")
		}
	}
	for _, field := range []string{"logical_message_id", "conversation_id", "reply_to_message_id", "text", "payload_b64u"} {
		if member, present := raw[field]; present {
			var text string
			if err := json.Unmarshal(member, &text); err != nil || text == "" {
				return invalidV2(field + " must be a non-empty string")
			}
		}
	}
	type applicationPlaintextAlias V2ApplicationPlaintext
	var decoded applicationPlaintextAlias
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&decoded); err != nil {
		return invalidV2(err.Error())
	}
	*value = V2ApplicationPlaintext(decoded)
	return nil
}

func (value V2ApplicationPlaintext) Validate() error {
	if value.ApplicationContentType == "" {
		return invalidV2("application_content_type must be non-empty")
	}
	count := 0
	if value.Text != nil {
		count++
	}
	if value.Payload != nil {
		count++
	}
	if value.PayloadB64U != nil {
		count++
	}
	if count != 1 {
		return invalidV2("exactly one plaintext bearer must be present")
	}
	if value.Text != nil && *value.Text == "" || value.PayloadB64U != nil && *value.PayloadB64U == "" {
		return invalidV2("plaintext bearer must be non-empty")
	}
	if value.PayloadB64U != nil {
		if err := validateB64U(*value.PayloadB64U, "payload_b64u"); err != nil {
			return err
		}
	}
	if value.ApplicationContentType == "text/plain" && value.Text == nil {
		return invalidV2("text/plain requires the text bearer")
	}
	if (value.ApplicationContentType == "application/json" || value.ApplicationContentType == "application/anp-attachment-manifest+json") && value.Payload == nil {
		return invalidV2(value.ApplicationContentType + " requires the payload bearer")
	}
	return nil
}

func decodeV2(value any, target any) error {
	encoded, err := json.Marshal(value)
	if err != nil {
		return err
	}
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(target); err != nil {
		return invalidV2(err.Error())
	}
	return nil
}

func empty(values ...string) bool {
	for _, value := range values {
		if value == "" {
			return true
		}
	}
	return false
}

func decimal(value string) bool {
	for _, char := range value {
		if char < '0' || char > '9' {
			return false
		}
	}
	return value != ""
}

func validateX25519B64U(value string, field string) error {
	if err := validateFixedB64U(value, field, 32); err != nil {
		return invalidV2(field + " must be unpadded base64url encoding a 32-byte X25519 public key")
	}
	return nil
}

func validateFixedB64U(value string, field string, expectedLength int) error {
	decoded, err := decodeV2B64U(value, field)
	if err != nil {
		return err
	}
	if len(decoded) != expectedLength {
		return invalidV2(field + " has the wrong decoded length")
	}
	return nil
}

func validateB64U(value string, field string) error {
	_, err := decodeV2B64U(value, field)
	return err
}

func decodeV2B64U(value string, field string) ([]byte, error) {
	if value == "" || strings.Contains(value, "=") {
		return nil, invalidV2(field + " must be unpadded base64url")
	}
	decoded, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		return nil, invalidV2(field + " must be base64url")
	}
	return decoded, nil
}

File: rust/src/direct_e2ee/v2_models.rs
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use chrono::DateTime;
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;

use super::v2_errors::DirectE2eeV2Error;

pub const DIRECT_E2EE_PROFILE_V2: &str = "anp.direct.e2ee.v2";
pub const DIRECT_E2EE_SECURITY_PROFILE: &str = "direct-e2ee";
pub const TRANSPORT_PROTECTED_SECURITY_PROFILE: &str = "transport-protected";
pub const CONTENT_TYPE_DIRECT_INIT_V2: &str = "application/anp-direct-init+json";
pub const CONTENT_TYPE_DIRECT_CIPHER_V2: &str = "application/anp-direct-cipher+json";
pub const MTI_DIRECT_E2EE_SUITE_V2: &str = "ANP-DIRECT-E2EE-X3DH-25519-CHACHA20POLY1305-SHA256-V1";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2SignedPrekey {
    pub key_id: String,
    pub public_key_b64u: String,
    pub expires_at: String,
}

impl V2SignedPrekey {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        require_non_empty("signed_prekey.key_id", &self.key_id)?;
        validate_x25519_b64u("signed_prekey.public_key_b64u", &self.public_key_b64u)?;
        DateTime::parse_from_rfc3339(&self.expires_at)
            .map_err(|_| DirectE2eeV2Error::invalid("signed_prekey.expires_at must be RFC3339"))?;
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2OneTimePrekey {
    pub key_id: String,
    pub public_key_b64u: String,
}

impl V2OneTimePrekey {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        require_non_empty("one_time_prekey.key_id", &self.key_id)?;
        validate_x25519_b64u("one_time_prekey.public_key_b64u", &self.public_key_b64u)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct V2PrekeyBundle {
    pub bundle_id: String,
    pub owner_did: String,
    pub owner_device_id: String,
    pub suite: String,
    pub static_key_agreement_id: String,
    pub signed_prekey: V2SignedPrekey,
    pub proof: Value,
}

impl V2PrekeyBundle {
    pub fn validate_structure(&self) -> Result<(), DirectE2eeV2Error> {
        for (name, value) in [
            ("prekey_bundle.bundle_id", self.bundle_id.as_str()),
            ("prekey_bundle.owner_did", self.owner_did.as_str()),
            (
                "prekey_bundle.owner_device_id",
                self.owner_device_id.as_str(),
            ),
            (
                "prekey_bundle.static_key_agreement_id",
                self.static_key_agreement_id.as_str(),
            ),
        ] {
            require_non_empty(name, value)?;
        }
        require_eq("prekey_bundle.suite", &self.suite, MTI_DIRECT_E2EE_SUITE_V2)?;
        self.signed_prekey.validate()?;
        let proof = self
            .proof
            .as_object()
            .ok_or_else(|| DirectE2eeV2Error::invalid("prekey_bundle.proof must be an object"))?;
        for field in [
            "type",
            "cryptosuite",
            "verificationMethod",
            "proofPurpose",
            "created",
            "proofValue",
        ] {
            if proof
                .get(field)
                .and_then(Value::as_str)
                .unwrap_or_default()
                .is_empty()
            {
                return Err(DirectE2eeV2Error::invalid(format!(
                    "prekey_bundle.proof.{field} must be a non-empty string"
                )));
            }
        }
        require_eq(
            "prekey_bundle.proof.type",
            proof["type"].as_str().expect("validated string"),
            "DataIntegrityProof",
        )?;
        require_eq(
            "prekey_bundle.proof.cryptosuite",
            proof["cryptosuite"].as_str().expect("validated string"),
            "eddsa-jcs-2022",
        )?;
        require_eq(
            "prekey_bundle.proof.proofPurpose",
            proof["proofPurpose"].as_str().expect("validated string"),
            "assertionMethod",
        )
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2Target {
    pub kind: String,
    pub did: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2KeyServiceMetadata {
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub anp_version: Option<String>,
    pub profile: String,
    pub security_profile: String,
    pub sender_did: String,
    pub sender_device_id: String,
    pub target: V2Target,
    pub operation_id: String,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub created_at: Option<String>,
}

impl V2KeyServiceMetadata {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        require_eq("meta.profile", &self.profile, DIRECT_E2EE_PROFILE_V2)?;
        require_eq(
            "meta.security_profile",
            &self.security_profile,
            TRANSPORT_PROTECTED_SECURITY_PROFILE,
        )?;
        require_eq("meta.target.kind", &self.target.kind, "service")?;
        for (name, value) in [
            ("meta.sender_did", self.sender_did.as_str()),
            ("meta.sender_device_id", self.sender_device_id.as_str()),
            ("meta.target.did", self.target.did.as_str()),
            ("meta.operation_id", self.operation_id.as_str()),
        ] {
            require_non_empty(name, value)?;
        }
        validate_optional("meta.anp_version", self.anp_version.as_deref())?;
        validate_optional_rfc3339("meta.created_at", self.created_at.as_deref())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2DirectMetadata {
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub anp_version: Option<String>,
    pub profile: String,
    pub security_profile: String,
    pub sender_did: String,
    pub sender_device_id: String,
    pub target: V2Target,
    pub recipient_device_id: String,
    pub operation_id: String,
    pub message_id: String,
    pub content_type: String,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub created_at: Option<String>,
}

impl V2DirectMetadata {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        require_eq("meta.profile", &self.profile, DIRECT_E2EE_PROFILE_V2)?;
        require_eq(
            "meta.security_profile",
            &self.security_profile,
            DIRECT_E2EE_SECURITY_PROFILE,
        )?;
        require_eq("meta.target.kind", &self.target.kind, "agent")?;
        if !matches!(
            self.content_type.as_str(),
            CONTENT_TYPE_DIRECT_INIT_V2 | CONTENT_TYPE_DIRECT_CIPHER_V2
        ) {
            return Err(DirectE2eeV2Error::invalid(
                "meta.content_type is not an MTI P5 v2 wire object",
            ));
        }
        for (name, value) in [
            ("meta.sender_did", self.sender_did.as_str()),
            ("meta.sender_device_id", self.sender_device_id.as_str()),
            ("meta.target.did", self.target.did.as_str()),
            (
                "meta.recipient_device_id",
                self.recipient_device_id.as_str(),
            ),
            ("meta.operation_id", self.operation_id.as_str()),
            ("meta.message_id", self.message_id.as_str()),
        ] {
            require_non_empty(name, value)?;
        }
        if self.operation_id != self.message_id {
            return Err(DirectE2eeV2Error::invalid(
                "meta.operation_id must equal meta.message_id",
            ));
        }
        validate_optional("meta.anp_version", self.anp_version.as_deref())?;
        validate_optional_rfc3339("meta.created_at", self.created_at.as_deref())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2RatchetHeader {
    pub dh_pub_b64u: String,
    pub pn: String,
    pub n: String,
}

impl V2RatchetHeader {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        validate_x25519_b64u("ratchet_header.dh_pub_b64u", &self.dh_pub_b64u)?;
        validate_decimal("ratchet_header.pn", &self.pn)?;
        validate_decimal("ratchet_header.n", &self.n)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2DirectInitBody {
    pub session_id: String,
    pub suite: String,
    pub sender_static_key_agreement_id: String,
    pub recipient_bundle_id: String,
    pub recipient_signed_prekey_id: String,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub recipient_one_time_prekey_id: Option<String>,
    pub sender_ephemeral_pub_b64u: String,
    pub ciphertext_b64u: String,
}

impl V2DirectInitBody {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        require_eq("body.suite", &self.suite, MTI_DIRECT_E2EE_SUITE_V2)?;
        validate_fixed_b64u("body.session_id", &self.session_id, 16)?;
        validate_x25519_b64u(
            "body.sender_ephemeral_pub_b64u",
            &self.sender_ephemeral_pub_b64u,
        )?;
        validate_b64u("body.ciphertext_b64u", &self.ciphertext_b64u)?;
        for (name, value) in [
            (
                "body.sender_static_key_agreement_id",
                self.sender_static_key_agreement_id.as_str(),
            ),
            (
                "body.recipient_bundle_id",
                self.recipient_bundle_id.as_str(),
            ),
            (
                "body.recipient_signed_prekey_id",
                self.recipient_signed_prekey_id.as_str(),
            ),
        ] {
            require_non_empty(name, value)?;
        }
        validate_optional(
            "body.recipient_one_time_prekey_id",
            self.recipient_one_time_prekey_id.as_deref(),
        )
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct V2DirectCipherBody {
    pub session_id: String,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub suite: Option<String>,
    pub ratchet_header: V2RatchetHeader,
    pub ciphertext_b64u: String,
}

impl V2DirectCipherBody {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        validate_fixed_b64u("body.session_id", &self.session_id, 16)?;
        validate_b64u("body.ciphertext_b64u", &self.ciphertext_b64u)?;
        if let Some(suite) = self.suite.as_deref() {
            require_eq("body.suite", suite, MTI_DIRECT_E2EE_SUITE_V2)?;
        }
        self.ratchet_header.validate()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct V2ApplicationPlaintext {
    pub application_content_type: String,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub logical_message_id: Option<String>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub conversation_id: Option<String>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub reply_to_message_id: Option<String>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub annotations: Option<Value>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub text: Option<String>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub payload: Option<Value>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_present"
    )]
    pub payload_b64u: Option<String>,
}

impl V2ApplicationPlaintext {
    pub fn validate(&self) -> Result<(), DirectE2eeV2Error> {
        require_non_empty("application_content_type", &self.application_content_type)?;
        for (name, value) in [
            ("logical_message_id", self.logical_message_id.as_deref()),
            ("conversation_id", self.conversation_id.as_deref()),
            ("reply_to_message_id", self.reply_to_message_id.as_deref()),
        ] {
            validate_optional(name, value)?;
        }
        let bearer_count = usize::from(self.text.is_some())
            + usize::from(self.payload.is_some())
            + usize::from(self.payload_b64u.is_some());
        if bearer_count != 1 {
            return Err(DirectE2eeV2Error::invalid(
                "exactly one of text, payload, or payload_b64u must be present",
            ));
        }
        if let Some(payload) = self.payload.as_ref() {
            if !payload.is_object() {
                return Err(DirectE2eeV2Error::invalid("payload must be an object"));
            }
        }
        if let Some(annotations) = self.annotations.as_ref() {
            if !annotations.is_object() {
                return Err(DirectE2eeV2Error::invalid("annotations must be an object"));
            }
        }
        if self.application_content_type == "text/plain" && self.text.is_none() {
            return Err(DirectE2eeV2Error::invalid(
                "text/plain requires the text bearer",
            ));
        }
        if matches!(
            self.application_content_type.as_str(),
            "application/json" | "application/anp-attachment-manifest+json"
        ) && self.payload.is_none()
        {
            return Err(DirectE2eeV2Error::invalid(format!(
                "{} requires the payload bearer",
                self.application_content_type
            )));
        }
        validate_optional("text", self.text.as_deref())?;
        if let Some(payload_b64u) = self.payload_b64u.as_deref() {
            validate_b64u("payload_b64u", payload_b64u)?;
        }
        Ok(())
    }
}

/// Deserialize an optional wire member while rejecting an explicit JSON null.
///
/// With `default`, serde calls this only when the member is present; the inner
/// type therefore owns null rejection instead of collapsing null into None.
pub(crate) fn deserialize_present<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
    D: Deserializer<'de>,
    T: Deserialize<'de>,
{
    T::deserialize(deserializer).map(Some)
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum V2DirectBody {
    Init(V2DirectInitBody),
    Cipher(V2DirectCipherBody),
}

pub(crate) fn require_non_empty(field: &str, value: &str) -> Result<(), DirectE2eeV2Error> {
    if value.is_empty() {
        Err(DirectE2eeV2Error::invalid(format!(
            "{field} must be a non-empty string"
        )))
    } else {
        Ok(())
    }
}

pub(crate) fn require_eq(
    field: &str,
    actual: &str,
    expected: &str,
) -> Result<(), DirectE2eeV2Error> {
    if actual == expected {
        Ok(())
    } else {
        Err(DirectE2eeV2Error::invalid(format!(
            "{field} must equal {expected}"
        )))
    }
}

fn validate_optional(field: &str, value: Option<&str>) -> Result<(), DirectE2eeV2Error> {
    if value.is_some_and(str::is_empty) {
        Err(DirectE2eeV2Error::invalid(format!(
            "{field} must be omitted rather than empty"
        )))
    } else {
        Ok(())
    }
}

fn validate_optional_rfc3339(field: &str, value: Option<&str>) -> Result<(), DirectE2eeV2Error> {
    if let Some(value) = value {
        require_non_empty(field, value)?;
        DateTime::parse_from_rfc3339(value)
            .map_err(|_| DirectE2eeV2Error::invalid(format!("{field} must be RFC3339")))?;
    }
    Ok(())
}

fn validate_decimal(field: &str, value: &str) -> Result<(), DirectE2eeV2Error> {
    if value.is_empty() || !value.bytes().all(|byte| byte.is_ascii_digit()) {
        return Err(DirectE2eeV2Error::invalid(format!(
            "{field} must be a decimal string"
        )));
    }
    Ok(())
}

fn validate_x25519_b64u(field: &str, value: &str) -> Result<(), DirectE2eeV2Error> {
    validate_fixed_b64u(field, value, 32).map_err(|_| {
        DirectE2eeV2Error::invalid(format!(
            "{field} must be unpadded base64url encoding a 32-byte X25519 public key"
        ))
    })
}

fn validate_fixed_b64u(
    field: &str,
    value: &str,
    expected_len: usize,
) -> Result<(), DirectE2eeV2Error> {
    let decoded = decode_b64u(field, value)?;
    if decoded.len() != expected_len {
        return Err(DirectE2eeV2Error::invalid(format!(
            "{field} must encode {expected_len} bytes"
        )));
    }
    Ok(())
}

fn validate_b64u(field: &str, value: &str) -> Result<(), DirectE2eeV2Error> {
    decode_b64u(field, value).map(|_| ())
}

fn decode_b64u(field: &str, value: &str) -> Result<Vec<u8>, DirectE2eeV2Error> {
    if value.is_empty() || value.contains('=') {
        return Err(DirectE2eeV2Error::invalid(format!(
            "{field} must be unpadded base64url"
        )));
    }
    URL_SAFE_NO_PAD
        .decode(value)
        .map_err(|_| DirectE2eeV2Error::invalid(format!("{field} must be base64url")))
}
